builtin/assign_osh.py

```diff
@@ -247,10 +247,16 @@ def _AssignVarForBuiltin(mem, rval, pair, which_scopes, flags, arith_ev,
         elif flag_A:
             if old_val.tag() in (value_e.Undef, value_e.Str,
                                  value_e.BashArray):
-                # Note: We explicitly initialize BashAssoc for Undef.
+                # Note: We explicitly initialize BashAssoc for Undef and Str.
+                #   When applying +=() to Str, we associate an old value to the
+                #   key '0'.
                 # Note: OSH allows overwriting an existing BashArray with an
                 #   empty BashAssoc.  Bash does not allow this.
-                val = bash_impl.BashAssoc_New()
+                assoc_val = bash_impl.BashAssoc_New()
+                if pair.plus_eq and old_val.tag() == value_e.Str:
+                    bash_impl.BashAssoc_SetElement(assoc_val, '0',
```

Comment:
What do you think about hiding this behavior behind `shopt -s strict_array` ?

It is to prevent confusion over types.  It currently does things like this:



```
$ bin/osh -o strict_array -c 'a=(x y); echo $a'
  a=(x y); echo $a
                ^~
[ -c flag ]:1: fatal: Array 'a' can't be referred to as a scalar (without @ or *)


$ bin/osh -o strict_array -c 's=foo; echo ${s[@]}'
  s=foo; echo ${s[@]}
              ^~
[ -c flag ]:1: fatal: Can't index string with @

```

Reply:
OK, I added it in commit 59ad69491.